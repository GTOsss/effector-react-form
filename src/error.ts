import {useStoreMap} from 'effector-react';
import {ErrorsInline, FieldsInline, FieldState, FormState, Message, UseErrorParams} from '../index';
import {getIn} from './utils/object-manager';
import {initialFieldState} from './form';

const useError = <Values>({
  name,
  $values,
  $errorsInline,
  $outerErrorsInline,
  $form,
  $fieldsInline,
}: UseErrorParams) => {
  const value = useStoreMap<Values, any, [string]>({
    store: $values,
    keys: [name],
    fn: (values, [field]) => getIn(values, field) || null,
  });

  const innerError = useStoreMap<ErrorsInline, Message, [string]>({
    store: $errorsInline,
    keys: [name],
    fn: (errorsInline, [field]) => errorsInline[field] || null,
  });
  const outerError = useStoreMap<ErrorsInline, Message, [string]>({
    store: $outerErrorsInline,
    keys: [name],
    fn: (outerErrorsInline, [field]) => outerErrorsInline[field] || null,
  });
  const error = innerError || outerError;

  const fieldState = useStoreMap<FieldsInline, FieldState, [string]>({
    store: $fieldsInline,
    keys: [name],
    fn: (fieldsInline, [field]) => fieldsInline[field] || initialFieldState,
  });

  const formSubmitted = useStoreMap<FormState, boolean, []>({
    store: $form,
    keys: [],
    fn: (formState) => formState.submitted,
  });
  const formHasError = useStoreMap<FormState, boolean, []>({
    store: $form,
    keys: [],
    fn: (formState) => formState.hasError,
  });
  const formHasOuterError = useStoreMap<FormState, boolean, []>({
    store: $form,
    keys: [],
    fn: (formState) => formState.hasOuterError,
  });
  const formState: FormState = {
    submitted: formSubmitted,
    hasError: formHasError,
    hasOuterError: formHasOuterError,
  };

  const isShowInnerError = (formState.submitted || fieldState.blurred) && Boolean(innerError);
  const isShowOuterError = (!fieldState.changedAfterOuterError) && Boolean(outerError);
  const isShowError = isShowInnerError || isShowOuterError;

  return {
    inputValue: value,
    form: formState,
    fieldState,
    error,
    innerError,
    outerError,
    isShowError,
    isShowOuterError,
    isShowInnerError,
  };
};

export default useError;
