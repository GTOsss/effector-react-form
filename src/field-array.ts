import React, {useCallback, useEffect, useMemo} from 'react';
import {useStore} from 'effector-react';
import {
  FieldArrayParams, MapFieldArrayCallback,
} from '../index';
import {getIn, setIn, removeFromInlineMap} from './utils/object-manager';
import {createEvent} from 'effector';

const useFieldArray = <Values>({$fieldsInline, $values, name}: FieldArrayParams<Values>) => {
  const remove = useMemo(() => createEvent<number>('hookForm_fieldArray_Remove'), []);
  const push = useMemo(() => createEvent<any>('hookForm_fieldArray_Push'), []);
  const unshift = useMemo(() => createEvent<any>('hookForm_fieldArray_Unshift'), []);
  const move = useMemo(() => createEvent<{from: number, to: number}>('hookForm_fieldArray_Move'), []);
  const insert = useMemo(() => createEvent<{value: any, index: number}>('hookForm_fieldArray_Insert'), []);
  const swap = useMemo(() => createEvent<{from: number, to: number}>('hookForm_fieldArray_Swap'), []);

  const values = useStore($values);

  const map = useCallback((callback: MapFieldArrayCallback) => {
    const results = [];
    const fields = getIn(values, name, []);
    fields.forEach((field, index) => {
      const callbackResult = callback({
        formItemName: `${name}.${index}`,
        fields,
        field,
        index,
      });
      results.push(callbackResult);
    });
    return results;
  }, [values]);

  useEffect(() => {
    $fieldsInline.on(remove, (fieldsInline, index) => removeFromInlineMap(fieldsInline, `${name}.${index}`));
    $values.on(remove, (values, index) => {
      const newFields = getIn(values, name, []).filter((_, i) => i !== index);
      return setIn(values, name, newFields);
    });

    // $fieldsInline will be initialize automatically because it's new field
    $values.on(push, (values, value) => {
      const newFields = [...getIn(values, name, []), value];
      return setIn(values, name, newFields);
    });

    $values.on(unshift, (values, value) => { // todo implement for fieldsInline
      const newFields = [value, ...getIn(values, name, [])];
      return setIn(values, name, newFields);
    });

    $values.on(move, (values, {from, to}) => { // todo implement for fieldsInline
      const fields = getIn(values, name, []);
      const newFields = [];
      let movingField = {};
      fields.forEach((field, i) => {
        if (from === i) {
          movingField = field;
        } else if (to === i) {
          newFields.push(field);
          newFields.push(movingField);
        } else {
          newFields.push(field);
        }
      });
      return setIn(values, name, newFields);
    });

    $values.on(insert, (values, {index, value}) => { // todo implement for fieldsInline
      const fields = getIn(values, name);
      const newFields = [];
      fields.forEach((field, i) => {
        if (index === i) {
          newFields.push(value);
          newFields.push(field);
        } else {
          newFields.push(field);
        }
      });
      return setIn(values, name, newFields);
    });

    $values.on(swap, (values, {from, to}) => { // todo implement for fieldsInline
      const fields = getIn(values, name);
      const newFields = [];
      fields.forEach((field, i) => {
        if (from === i) {
          newFields.push(fields[to]);
        } else if (to === i) {
          newFields.push(fields[from]);
        } else {
          newFields.push(field);
        }
      });
      return setIn(values, name, newFields);
    });

    return () => {
      $values.off(remove);
      $values.off(push);
      $values.off(unshift);
      $values.off(move);
      $values.off(insert);
      $values.off(swap);
    };
  }, []);

  return {
    map,
    remove,
    push,
    unshift,
    move,
    insert,
    swap,
  };
};

export default useFieldArray;
