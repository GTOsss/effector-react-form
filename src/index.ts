import useForm from './use-form';
import useFieldArray from './use-field-array';
import useError from './use-error';
import createFieldArray from './factories/create-field-array';
import createForm from './factories/create-form';
import { deleteIn, getIn, setIn, makeNested, removeFromInlineMap } from './utils/object-manager';

export * from './ts';
export {
  createForm,
  createFieldArray,
  useForm,
  useError,
  useFieldArray,
  deleteIn,
  getIn,
  setIn,
  makeNested,
  removeFromInlineMap,
};
