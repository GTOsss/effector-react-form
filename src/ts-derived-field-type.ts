type IsRecordLike<T> = T extends object ? (string extends keyof T ? true : false) : false;

export type DerivedFieldType<T> = IsRecordLike<T> extends true
  ? string
  : T extends object
  ? {
      [K in keyof T]: K extends string | number
        ? T[K] extends object
          ? K | `${K}.${DerivedFieldType<T[K]>}`
          : K
        : never;
    }[keyof T]
  : never;
