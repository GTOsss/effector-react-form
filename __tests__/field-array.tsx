import { Controller } from '../src/ts';
import React from 'react';
import { useForm, useFieldArray } from '../src';
import { render, screen, fireEvent } from '@testing-library/react';
import { createEvent } from 'effector';
import createForm from '../src/factories/create-form';
import createFieldArray from '../src/factories/create-field-array';

let counterGlobal = 0;
const getId = () => (counterGlobal += 1);

const reset = createEvent();

reset.watch(() => (counterGlobal = 0));

interface InputProps {
  controller: Controller;
  label: string;
}

const validateRequired = (value) => (value ? undefined : 'Field is required');

const Input: React.FC<InputProps> = ({ controller, label }) => {
  const { input, fieldState, form, error } = controller();

  const isShowError = error && (form.submitted || form.hasOuterError || fieldState.blurred);

  return (
    <div role="wrapper-for-input" className="input-wrap">
      <label>{label}</label>
      <input
        {...input}
        value={input.value || ''}
        role="textbox"
        className={`input${isShowError ? ' input-error' : ''}`}
        placeholder={label}
      />
      {isShowError && <span>{error}</span>}
    </div>
  );
};

const FormItemsInner = ({ controller, name, onPushInner, fieldArray }) => {
  const { map, push, remove } = useFieldArray({ name, fieldArray });

  return (
    <div className="formItemsInner" role="formItemsInner">
      {map(({ formItemName, index, field }) => (
        <div key={field.id} className="formItem" role="formItemInner">
          <Input
            label="Username"
            controller={controller({ name: `${formItemName}.username`, validate: validateRequired })}
          />
          <Input
            label="First name"
            controller={controller({ name: `${formItemName}.profile.firstName`, validate: validateRequired })}
          />
          <button type="button" onClick={() => remove(index)}>
            remove form item inner
          </button>
        </div>
      ))}
      <button type="button" onClick={() => onPushInner(push)}>
        add inner form item
      </button>
    </div>
  );
};

const FormItems = ({ controller, name, onPush, onPushInner, fieldArray }) => {
  const { map, push, remove } = useFieldArray({ name, fieldArray });

  return (
    <div className="formItems" role="formItems">
      {map(({ formItemName, index, field }) => (
        <div key={field.id} className="formItem" role="formItem">
          <Input
            label="Username"
            controller={controller({ name: `${formItemName}.username`, validate: validateRequired })}
          />
          <Input
            label="First name"
            controller={controller({ name: `${formItemName}.profile.firstName`, validate: validateRequired })}
          />

          <FormItemsInner
            controller={controller}
            name={`${formItemName}.inners`}
            onPushInner={onPushInner}
            fieldArray={fieldArray}
          />
          <button type="button" onClick={() => remove(index)}>
            remove form item
          </button>
        </div>
      ))}
      <button type="button" onClick={() => onPush(push)}>
        add form item with values
      </button>
    </div>
  );
};

const FieldArray = ({ form, fieldArray, onPush, onPushInner }: any) => {
  const { handleSubmit, controller } = useForm({ form: form });

  return (
    <form onSubmit={handleSubmit}>
      <FormItems
        name="parents"
        fieldArray={fieldArray}
        controller={controller}
        onPush={onPush}
        onPushInner={onPushInner}
      />
      <button type="submit">submit</button>
    </form>
  );
};

describe('FieldArray', () => {
  test('render', () => {
    reset();
    const form = createForm();
    const fieldArray = createFieldArray({ form });
    render(<FieldArray form={form} fieldArray={fieldArray} onPush={() => null} />);
    const formItems = screen.getAllByRole('formItems');
    expect(formItems).toMatchSnapshot();
  });

  test('add form item with values > render', () => {
    reset();
    const form = createForm();
    const fieldArray = createFieldArray({ form });
    render(
      <FieldArray
        form={form}
        fieldArray={fieldArray}
        onPush={(push) => push({ id: getId(), username: 'gto', profile: { firstName: 'Timofey' } })}
      />,
    );
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    const formItems = screen.getAllByRole('formItems');
    expect(formItems).toMatchSnapshot();
  });

  test('add form item with values > $values', () => {
    reset();
    const form = createForm();
    const fieldArray = createFieldArray({ form });
    render(
      <FieldArray
        form={form}
        fieldArray={fieldArray}
        onPush={(push) => push({ id: getId(), username: 'gto', profile: { firstName: 'Timofey' } })}
      />,
    );
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    expect(form.$values.getState()).toMatchSnapshot();
  });

  test('add form item with values (x3) > $values', () => {
    reset();
    const formItems = [
      { id: getId(), username: '0' },
      { id: getId(), username: '1' },
      { id: getId(), username: '2' },
    ];
    const form = createForm();
    const fieldArray = createFieldArray({ form });

    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };
    render(<FieldArray form={form} fieldArray={fieldArray} onPush={onPush} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    expect(form.$values.getState()).toMatchSnapshot();
  });

  test('add form item with values (x3) > remove(1) > $values', () => {
    reset();
    const formItems = [
      { id: getId(), username: '0' },
      { id: getId(), username: '1' },
      { id: getId(), username: '2' },
    ];
    const form = createForm();
    const fieldArray = createFieldArray({ form });

    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };
    render(<FieldArray form={form} fieldArray={fieldArray} onPush={onPush} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    const removeFormItem = screen.getAllByText('remove form item');
    fireEvent.click(removeFormItem[1]);
    expect(form.$values.getState()).toMatchSnapshot();
  });

  test('add form item with values (x3) > remove(1) > $fieldsInline', () => {
    reset();
    const formItems = [
      { id: getId(), username: '0' },
      { id: getId(), username: '1' },
      { id: getId(), username: '2' },
    ];
    const form = createForm();
    const fieldArray = createFieldArray({ form });

    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };
    render(<FieldArray form={form} fieldArray={fieldArray} onPush={onPush} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    const removeFormItem = screen.getAllByText('remove form item');
    fireEvent.click(removeFormItem[1]);
    expect(form.$fieldsInline.getState()).toMatchSnapshot();
  });

  test('add form item with values (x3) > add inner (x3) > $values', () => {
    reset();
    const formItems = [
      { id: getId(), username: '0' },
      { id: getId(), username: '1' },
      { id: getId(), username: '2' },
    ];
    const form = createForm();
    const fieldArray = createFieldArray({ form });

    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };

    const formItemsInner = [
      { id: getId(), username: 'inner 0' },
      { id: getId(), username: 'inner 1' },
      { id: getId(), username: 'inner 2' },
    ];
    let counterInner = 0;
    const onPushInner = (push) => {
      push(formItemsInner[counterInner++]);
    };

    render(<FieldArray form={form} fieldArray={fieldArray} onPush={onPush} onPushInner={onPushInner} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    const addFormItemInner = screen.getAllByText('add inner form item');
    fireEvent.click(addFormItemInner[0]);
    fireEvent.click(addFormItemInner[1]);
    fireEvent.click(addFormItemInner[2]);
    expect(form.$values.getState()).toMatchSnapshot();
  });

  test('add form item with values (x3) > add inner (x3) > render', () => {
    reset();
    const formItems = [
      { id: getId(), username: '0' },
      { id: getId(), username: '1' },
      { id: getId(), username: '2' },
    ];
    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };

    const formItemsInner = [
      { id: getId(), username: 'inner 0' },
      { id: getId(), username: 'inner 1' },
      { id: getId(), username: 'inner 2' },
    ];
    let counterInner = 0;
    const onPushInner = (push) => {
      push(formItemsInner[counterInner++]);
    };

    const form = createForm();
    const fieldArray = createFieldArray({ form });
    render(<FieldArray form={form} fieldArray={fieldArray} onPush={onPush} onPushInner={onPushInner} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    const addFormItemInner = screen.getAllByText('add inner form item');
    fireEvent.click(addFormItemInner[0]);
    fireEvent.click(addFormItemInner[1]);
    fireEvent.click(addFormItemInner[2]);
    const formItemsDom = screen.getByRole('formItems');
    expect(formItemsDom).toMatchSnapshot();
  });

  test('add form item with values (x3) > add inner (x3) > remove form item 1 > $values', () => {
    reset();
    const formItems = [
      { id: getId(), username: '0' },
      { id: getId(), username: '1' },
      { id: getId(), username: '2' },
    ];
    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };

    const formItemsInner = [
      { id: getId(), username: 'inner 0' },
      { id: getId(), username: 'inner 1' },
      { id: getId(), username: 'inner 2' },
    ];
    let counterInner = 0;
    const onPushInner = (push) => {
      push(formItemsInner[counterInner++]);
    };

    const form = createForm();
    const fieldArray = createFieldArray({ form });
    render(<FieldArray form={form} fieldArray={fieldArray} onPush={onPush} onPushInner={onPushInner} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    const addFormItemInner = screen.getAllByText('add inner form item');
    fireEvent.click(addFormItemInner[0]);
    fireEvent.click(addFormItemInner[1]);
    fireEvent.click(addFormItemInner[2]);
    const removeFormItem = screen.getAllByText('remove form item');
    fireEvent.click(removeFormItem[1]);
    expect(form.$values.getState()).toMatchSnapshot();
  });

  test('add form item with values (x3) > add inner (x3) > remove form item 1 > render', () => {
    reset();
    const formItems = [
      { id: getId(), username: '0' },
      { id: getId(), username: '1' },
      { id: getId(), username: '2' },
    ];
    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };

    const formItemsInner = [
      { id: getId(), username: 'inner 0' },
      { id: getId(), username: 'inner 1' },
      { id: getId(), username: 'inner 2' },
    ];
    let counterInner = 0;
    const onPushInner = (push) => {
      push(formItemsInner[counterInner++]);
    };

    const form = createForm();
    const fieldArray = createFieldArray({ form });
    render(<FieldArray form={form} fieldArray={fieldArray} onPush={onPush} onPushInner={onPushInner} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    const addFormItemInner = screen.getAllByText('add inner form item');
    fireEvent.click(addFormItemInner[0]);
    fireEvent.click(addFormItemInner[1]);
    fireEvent.click(addFormItemInner[2]);
    const removeFormItem = screen.getAllByText('remove form item');
    fireEvent.click(removeFormItem[1]);
    const formItemsDom = screen.getByRole('formItems');
    expect(formItemsDom).toMatchSnapshot();
  });

  test('add form item with values (x2) > remove 1 > add inner 0 (x1) > $values', () => {
    reset();
    const formItems = [
      { id: getId(), username: '0' },
      { id: getId(), username: '1' },
    ];
    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };

    const formItemsInner = [
      { id: getId(), username: 'inner 0' },
      { id: getId(), username: 'inner 1' },
    ];
    let counterInner = 0;
    const onPushInner = (push) => {
      push(formItemsInner[counterInner++]);
    };

    const form = createForm();
    const fieldArray = createFieldArray({ form });
    render(<FieldArray form={form} fieldArray={fieldArray} onPush={onPush} onPushInner={onPushInner} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    const removeFormItem = screen.getAllByText('remove form item');
    fireEvent.click(removeFormItem[0]);
    const addFormItemInner = screen.getAllByText('add inner form item');
    fireEvent.click(addFormItemInner[0]);
    expect(form.$values.getState()).toMatchSnapshot();
  });

  test('add form item with values (x2) > remove 1 > add inner 0 (x1) > render', () => {
    reset();
    const formItems = [
      { id: getId(), username: '0' },
      { id: getId(), username: '1' },
    ];
    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };

    const formItemsInner = [
      { id: getId(), username: 'inner 0' },
      { id: getId(), username: 'inner 1' },
    ];
    let counterInner = 0;
    const onPushInner = (push) => {
      push(formItemsInner[counterInner++]);
    };

    const form = createForm();
    const fieldArray = createFieldArray({ form });
    render(<FieldArray form={form} fieldArray={fieldArray} onPush={onPush} onPushInner={onPushInner} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    const removeFormItem = screen.getAllByText('remove form item');
    fireEvent.click(removeFormItem[0]);
    const addFormItemInner = screen.getAllByText('add inner form item');
    fireEvent.click(addFormItemInner[0]);
    const formItemsDom = screen.getByRole('formItems');
    expect(formItemsDom).toMatchSnapshot();
  });
});
