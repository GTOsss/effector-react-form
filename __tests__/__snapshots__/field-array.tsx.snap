// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldArray add form item with values (x2) > remove 1 > add inner 0 (x1) > $values 1`] = `
Object {
  "parents": Array [
    Object {
      "id": 2,
      "inners": Array [
        Object {
          "id": 3,
          "username": "inner 0",
        },
      ],
      "username": "1",
    },
  ],
}
`;

exports[`FieldArray add form item with values (x2) > remove 1 > add inner 0 (x1) > render 1`] = `
<div
  class="formItems"
  role="formItems"
>
  <div
    class="formItem"
    role="formItem"
  >
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        Username
      </label>
      <input
        class="input"
        name="parents.0.username"
        placeholder="Username"
        role="textbox"
        value="1"
      />
    </div>
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        First name
      </label>
      <input
        class="input"
        name="parents.0.profile.firstName"
        placeholder="First name"
        role="textbox"
        value=""
      />
    </div>
    <div
      class="formItemsInner"
      role="formItemsInner"
    >
      <div
        class="formItem"
        role="formItemInner"
      >
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            Username
          </label>
          <input
            class="input"
            name="parents.0.inners.0.username"
            placeholder="Username"
            role="textbox"
            value="inner 0"
          />
        </div>
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            First name
          </label>
          <input
            class="input"
            name="parents.0.inners.0.profile.firstName"
            placeholder="First name"
            role="textbox"
            value=""
          />
        </div>
        <button
          type="button"
        >
          remove form item inner
        </button>
      </div>
      <button
        type="button"
      >
        add inner form item
      </button>
    </div>
    <button
      type="button"
    >
      remove form item
    </button>
  </div>
  <button
    type="button"
  >
    add form item with values
  </button>
</div>
`;

exports[`FieldArray add form item with values (x3) > $values 1`] = `
Object {
  "parents": Array [
    Object {
      "id": 1,
      "username": "0",
    },
    Object {
      "id": 2,
      "username": "1",
    },
    Object {
      "id": 3,
      "username": "2",
    },
  ],
}
`;

exports[`FieldArray add form item with values (x3) > add inner (x3) > $values 1`] = `
Object {
  "parents": Array [
    Object {
      "id": 1,
      "inners": Array [
        Object {
          "id": 4,
          "username": "inner 0",
        },
      ],
      "username": "0",
    },
    Object {
      "id": 2,
      "inners": Array [
        Object {
          "id": 5,
          "username": "inner 1",
        },
      ],
      "username": "1",
    },
    Object {
      "id": 3,
      "inners": Array [
        Object {
          "id": 6,
          "username": "inner 2",
        },
      ],
      "username": "2",
    },
  ],
}
`;

exports[`FieldArray add form item with values (x3) > add inner (x3) > remove form item 1 > $values 1`] = `
Object {
  "parents": Array [
    Object {
      "id": 1,
      "inners": Array [
        Object {
          "id": 4,
          "username": "inner 0",
        },
      ],
      "username": "0",
    },
    Object {
      "id": 3,
      "inners": Array [
        Object {
          "id": 6,
          "username": "inner 2",
        },
      ],
      "username": "2",
    },
  ],
}
`;

exports[`FieldArray add form item with values (x3) > add inner (x3) > remove form item 1 > render 1`] = `
<div
  class="formItems"
  role="formItems"
>
  <div
    class="formItem"
    role="formItem"
  >
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        Username
      </label>
      <input
        class="input"
        name="parents.0.username"
        placeholder="Username"
        role="textbox"
        value="0"
      />
    </div>
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        First name
      </label>
      <input
        class="input"
        name="parents.0.profile.firstName"
        placeholder="First name"
        role="textbox"
        value=""
      />
    </div>
    <div
      class="formItemsInner"
      role="formItemsInner"
    >
      <div
        class="formItem"
        role="formItemInner"
      >
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            Username
          </label>
          <input
            class="input"
            name="parents.0.inners.0.username"
            placeholder="Username"
            role="textbox"
            value="inner 0"
          />
        </div>
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            First name
          </label>
          <input
            class="input"
            name="parents.0.inners.0.profile.firstName"
            placeholder="First name"
            role="textbox"
            value=""
          />
        </div>
        <button
          type="button"
        >
          remove form item inner
        </button>
      </div>
      <button
        type="button"
      >
        add inner form item
      </button>
    </div>
    <button
      type="button"
    >
      remove form item
    </button>
  </div>
  <div
    class="formItem"
    role="formItem"
  >
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        Username
      </label>
      <input
        class="input"
        name="parents.1.username"
        placeholder="Username"
        role="textbox"
        value="2"
      />
    </div>
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        First name
      </label>
      <input
        class="input"
        name="parents.1.profile.firstName"
        placeholder="First name"
        role="textbox"
        value=""
      />
    </div>
    <div
      class="formItemsInner"
      role="formItemsInner"
    >
      <div
        class="formItem"
        role="formItemInner"
      >
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            Username
          </label>
          <input
            class="input"
            name="parents.1.inners.0.username"
            placeholder="Username"
            role="textbox"
            value="inner 2"
          />
        </div>
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            First name
          </label>
          <input
            class="input"
            name="parents.1.inners.0.profile.firstName"
            placeholder="First name"
            role="textbox"
            value=""
          />
        </div>
        <button
          type="button"
        >
          remove form item inner
        </button>
      </div>
      <button
        type="button"
      >
        add inner form item
      </button>
    </div>
    <button
      type="button"
    >
      remove form item
    </button>
  </div>
  <button
    type="button"
  >
    add form item with values
  </button>
</div>
`;

exports[`FieldArray add form item with values (x3) > add inner (x3) > render 1`] = `
<div
  class="formItems"
  role="formItems"
>
  <div
    class="formItem"
    role="formItem"
  >
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        Username
      </label>
      <input
        class="input"
        name="parents.0.username"
        placeholder="Username"
        role="textbox"
        value="0"
      />
    </div>
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        First name
      </label>
      <input
        class="input"
        name="parents.0.profile.firstName"
        placeholder="First name"
        role="textbox"
        value=""
      />
    </div>
    <div
      class="formItemsInner"
      role="formItemsInner"
    >
      <div
        class="formItem"
        role="formItemInner"
      >
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            Username
          </label>
          <input
            class="input"
            name="parents.0.inners.0.username"
            placeholder="Username"
            role="textbox"
            value="inner 0"
          />
        </div>
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            First name
          </label>
          <input
            class="input"
            name="parents.0.inners.0.profile.firstName"
            placeholder="First name"
            role="textbox"
            value=""
          />
        </div>
        <button
          type="button"
        >
          remove form item inner
        </button>
      </div>
      <button
        type="button"
      >
        add inner form item
      </button>
    </div>
    <button
      type="button"
    >
      remove form item
    </button>
  </div>
  <div
    class="formItem"
    role="formItem"
  >
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        Username
      </label>
      <input
        class="input"
        name="parents.1.username"
        placeholder="Username"
        role="textbox"
        value="1"
      />
    </div>
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        First name
      </label>
      <input
        class="input"
        name="parents.1.profile.firstName"
        placeholder="First name"
        role="textbox"
        value=""
      />
    </div>
    <div
      class="formItemsInner"
      role="formItemsInner"
    >
      <div
        class="formItem"
        role="formItemInner"
      >
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            Username
          </label>
          <input
            class="input"
            name="parents.1.inners.0.username"
            placeholder="Username"
            role="textbox"
            value="inner 1"
          />
        </div>
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            First name
          </label>
          <input
            class="input"
            name="parents.1.inners.0.profile.firstName"
            placeholder="First name"
            role="textbox"
            value=""
          />
        </div>
        <button
          type="button"
        >
          remove form item inner
        </button>
      </div>
      <button
        type="button"
      >
        add inner form item
      </button>
    </div>
    <button
      type="button"
    >
      remove form item
    </button>
  </div>
  <div
    class="formItem"
    role="formItem"
  >
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        Username
      </label>
      <input
        class="input"
        name="parents.2.username"
        placeholder="Username"
        role="textbox"
        value="2"
      />
    </div>
    <div
      class="input-wrap"
      role="wrapper-for-input"
    >
      <label>
        First name
      </label>
      <input
        class="input"
        name="parents.2.profile.firstName"
        placeholder="First name"
        role="textbox"
        value=""
      />
    </div>
    <div
      class="formItemsInner"
      role="formItemsInner"
    >
      <div
        class="formItem"
        role="formItemInner"
      >
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            Username
          </label>
          <input
            class="input"
            name="parents.2.inners.0.username"
            placeholder="Username"
            role="textbox"
            value="inner 2"
          />
        </div>
        <div
          class="input-wrap"
          role="wrapper-for-input"
        >
          <label>
            First name
          </label>
          <input
            class="input"
            name="parents.2.inners.0.profile.firstName"
            placeholder="First name"
            role="textbox"
            value=""
          />
        </div>
        <button
          type="button"
        >
          remove form item inner
        </button>
      </div>
      <button
        type="button"
      >
        add inner form item
      </button>
    </div>
    <button
      type="button"
    >
      remove form item
    </button>
  </div>
  <button
    type="button"
  >
    add form item with values
  </button>
</div>
`;

exports[`FieldArray add form item with values (x3) > remove(1) > $fieldsInline 1`] = `
Object {
  "parents.0.profile.firstName": Object {
    "_type": "fieldMeta",
    "active": false,
    "blurred": false,
    "blurredAfterOuterError": false,
    "changed": false,
    "changedAfterOuterError": false,
    "touched": false,
    "touchedAfterOuterError": false,
    "validate": [Function],
  },
  "parents.0.username": Object {
    "_type": "fieldMeta",
    "active": false,
    "blurred": false,
    "blurredAfterOuterError": false,
    "changed": false,
    "changedAfterOuterError": false,
    "touched": false,
    "touchedAfterOuterError": false,
    "validate": [Function],
  },
  "parents.1.profile.firstName": Object {
    "_type": "fieldMeta",
    "active": false,
    "blurred": false,
    "blurredAfterOuterError": false,
    "changed": false,
    "changedAfterOuterError": false,
    "touched": false,
    "touchedAfterOuterError": false,
    "validate": [Function],
  },
  "parents.1.username": Object {
    "_type": "fieldMeta",
    "active": false,
    "blurred": false,
    "blurredAfterOuterError": false,
    "changed": false,
    "changedAfterOuterError": false,
    "touched": false,
    "touchedAfterOuterError": false,
    "validate": [Function],
  },
}
`;

exports[`FieldArray add form item with values (x3) > remove(1) > $values 1`] = `
Object {
  "parents": Array [
    Object {
      "id": 1,
      "username": "0",
    },
    Object {
      "id": 3,
      "username": "2",
    },
  ],
}
`;

exports[`FieldArray add form item with values > $values 1`] = `
Object {
  "parents": Array [
    Object {
      "id": 1,
      "profile": Object {
        "firstName": "Timofey",
      },
      "username": "gto",
    },
  ],
}
`;

exports[`FieldArray add form item with values > render 1`] = `
Array [
  <div
    class="formItems"
    role="formItems"
  >
    <div
      class="formItem"
      role="formItem"
    >
      <div
        class="input-wrap"
        role="wrapper-for-input"
      >
        <label>
          Username
        </label>
        <input
          class="input"
          name="parents.0.username"
          placeholder="Username"
          role="textbox"
          value="gto"
        />
      </div>
      <div
        class="input-wrap"
        role="wrapper-for-input"
      >
        <label>
          First name
        </label>
        <input
          class="input"
          name="parents.0.profile.firstName"
          placeholder="First name"
          role="textbox"
          value="Timofey"
        />
      </div>
      <div
        class="formItemsInner"
        role="formItemsInner"
      >
        <button
          type="button"
        >
          add inner form item
        </button>
      </div>
      <button
        type="button"
      >
        remove form item
      </button>
    </div>
    <button
      type="button"
    >
      add form item with values
    </button>
  </div>,
]
`;

exports[`FieldArray render 1`] = `
Array [
  <div
    class="formItems"
    role="formItems"
  >
    <button
      type="button"
    >
      add form item with values
    </button>
  </div>,
]
`;
